lib/Dialect/FIRRTL/Transforms/ModuleInliner.cpp

```diff
@@ -0,0 +1,394 @@
+      if (liveModules.insert(target).second) {
+        worklist.push_back(target);
+      }
+      cloneAndRename(prefix, b, mapper, op);
+      continue;
+    }
+
+    // Create the wire mapping for results + ports.
+    auto instanceName = instance.nameAttr().getValue();
+    auto nestedPrefix = (prefix + instanceName + "_").str();
+    auto wires = mapPortsToWires(nestedPrefix, b, mapper, target);
+    mapResultsToWires(mapper, wires, instance);
+
+    // Inline the module, it can be marked as flatten and inline.
+    if (shouldFlatten(target)) {
+      flattenInto(nestedPrefix, b, mapper, target);
+    } else {
+      inlineInto(nestedPrefix, b, mapper, target);
+    }
+  }
+}
+
+void Inliner::inlineInstances(FModuleOp module) {
+  for (auto &op : llvm::make_early_inc_range(*module.getBodyBlock())) {
+    // If its not an instance op, skip it.
+    auto instance = dyn_cast<InstanceOp>(op);
+    if (!instance)
+      continue;
+
+    // If its not a regular module we can't inline it. Mark is as live.
+    auto module = symbolTable.lookup(instance.moduleName());
+    auto target = dyn_cast<FModuleOp>(module);
+    if (!target) {
+      liveModules.insert(module);
+      continue;
+    }
+
+    // If we aren't inlining the target, add it to the work list.
+    if (!shouldInline(target)) {
+      if (liveModules.insert(target).second) {
+        worklist.push_back(target);
+      }
+      continue;
+    }
+
+    // Create the wire mapping for results + ports. We RAUW the results instead
+    // of mapping them.
+    BlockAndValueMapping mapper;
+    OpBuilder b(instance);
+    auto instanceName = instance.nameAttr().getValue();
+    auto nestedPrefix = (instanceName + "_").str();
+    auto wires = mapPortsToWires(nestedPrefix, b, mapper, target);
+    for (unsigned i = 0; i < instance.getNumResults(); ++i)
+      instance.getResult(i).replaceAllUsesWith(wires[i]);
+
+    // Inline the module, it can be marked as flatten and inline.
+    if (shouldFlatten(target)) {
+      flattenInto(nestedPrefix, b, mapper, target);
+    } else {
+      inlineInto(nestedPrefix, b, mapper, target);
+    }
+
+    // Erase the replaced instance.
+    instance.erase();
+  }
+}
+
+Inliner::Inliner(CircuitOp circuit)
+    : circuit(circuit), context(circuit.getContext()), symbolTable(circuit) {
+  // Saving these attributes here makes lookup faster later.
+  auto inlineId = NamedAttribute(
+      Identifier::get("class", context),
+      StringAttr::get(context, "firrtl.passes.InlineAnnotation"));
+  inlineAttr = DictionaryAttr::getWithSorted(context, {inlineId});
+  auto flattenId = NamedAttribute(
+      Identifier::get("class", context),
+      StringAttr::get(context, "firrtl.transforms.FlattenAnnotation"));
+  flattenAttr = DictionaryAttr::getWithSorted(context, {flattenId});
+}
+
+void Inliner::run() {
+  // If the top module is not a regular module, there is nothing to do.
+  auto topModule = dyn_cast<FModuleOp>(symbolTable.lookup(circuit.name()));
```

Comment:
CircuitOp has a helper for this, plz use it.

Reply:
Fixed!  I was worried that the lookup function would construct its own symbol table, but looks like it just does a linear scan through IR.